syntax = "proto3";

package dht;

// DHT Service Definition
service DHTService {
    // Client-facing operations
    // Put operation to store a key-value pair
    rpc Put(PutRequest) returns (PutResponse);
    
    // Get operation to retrieve a value by key
    rpc Get(GetRequest) returns (GetResponse);
    
    // Health check for node availability
    rpc HealthCheck(HealthCheckRequest) returns (HealthCheckResponse);
    
    // Internal operations (node-to-node)
    // Internal put for replication
    rpc InternalPut(InternalPutRequest) returns (InternalPutResponse);
    
    // Internal get for quorum reads
    rpc InternalGet(InternalGetRequest) returns (InternalGetResponse);
    
    // Read repair operation
    rpc ReadRepair(ReadRepairRequest) returns (ReadRepairResponse);
    
    // Get ring information (for node discovery)
    rpc GetRing(GetRingRequest) returns (GetRingResponse);
    
    // Join cluster (notify other nodes of our presence)
    rpc JoinCluster(JoinClusterRequest) returns (JoinClusterResponse);
    
    // Hinted Handoff operations
    // Internal put with hint (for sloppy quorum)
    rpc InternalPutWithHint(InternalPutWithHintRequest) returns (InternalPutWithHintResponse);
    
    // Deliver hint to target node
    rpc DeliverHint(DeliverHintRequest) returns (DeliverHintResponse);

    rpc InspectNode(InspectNodeRequest) returns (InspectNodeResponse);
}

// Put Request
message PutRequest {
    string key = 1;
    string value = 2;
    map<string, int32> vector_clock = 3;  // Vector clock for versioning (Phase 2)
}

// Put Response
message PutResponse {
    bool success = 1;
    string message = 2;
    map<string, int32> vector_clock = 3;
}

// Get Request
message GetRequest {
    string key = 1;
}

// Get Response
message GetResponse {
    bool success = 1;
    string key = 2;
    string value = 3;
    string message = 4;
    map<string, int32> vector_clock = 5;
    bool has_conflicts = 6;  // True if multiple concurrent versions exist
    repeated string sibling_values = 7;  // All sibling values in case of conflict
    repeated VectorClockMsg sibling_clocks = 8;  // Vector clocks for siblings
}

// Vector Clock Message (for repeated use)
message VectorClockMsg {
    map<string, int32> clock = 1;
}

// Versioned Data (for internal operations)
message VersionedData {
    string value = 1;
    map<string, int32> vector_clock = 2;
}

// Internal Put Request (for replication)
message InternalPutRequest {
    string key = 1;
    string value = 2;
    map<string, int32> vector_clock = 3;
    bool is_replication = 4;
}

// Internal Put Response
message InternalPutResponse {
    bool success = 1;
    string message = 2;
}

// Internal Get Request (for quorum reads)
message InternalGetRequest {
    string key = 1;
}

// Internal Get Response
message InternalGetResponse {
    bool success = 1;
    repeated VersionedData versions = 2;
}

// Read Repair Request
message ReadRepairRequest {
    string key = 1;
    repeated VersionedData versions = 2;
}

// Read Repair Response
message ReadRepairResponse {
    bool success = 1;
    string message = 2;
}

// Health Check Request
message HealthCheckRequest {
    string node_id = 1;
}

// Health Check Response
message HealthCheckResponse {
    bool healthy = 1;
    string node_id = 2;
    string datacenter = 3;  // Datacenter ID (Phase 3)
}

// Get Ring Request
message GetRingRequest {
    string requesting_node_id = 1;
}

// Get Ring Response
message GetRingResponse {
    repeated string nodes = 1;  // List of all nodes in the ring (host:port format)
    map<string, string> datacenter_map = 2;  // Map of node address to datacenter
}

// Join Cluster Request
message JoinClusterRequest {
    string node_address = 1;  // Address of joining node
    string datacenter = 2;     // Datacenter of joining node
}

// Join Cluster Response
message JoinClusterResponse {
    bool success = 1;
    string message = 2;
}

// Internal Put With Hint Request (for sloppy quorum)
message InternalPutWithHintRequest {
    string key = 1;
    string value = 2;
    map<string, int32> vector_clock = 3;
    string hint_for_node = 4;  // Address of node this data belongs to (empty if no hint)
}

// Internal Put With Hint Response
message InternalPutWithHintResponse {
    bool success = 1;
    string message = 2;
}

// Deliver Hint Request (transfer hinted data to recovered node)
message DeliverHintRequest {
    string key = 1;
    string value = 2;
    map<string, int32> vector_clock = 3;
}

// Deliver Hint Response
message DeliverHintResponse {
    bool success = 1;
    string message = 2;
}

// Add these messages at the bottom
message InspectNodeRequest { }

message InspectNodeResponse {
    bool success = 1;
    repeated StorageEntry entries = 2;
    string datacenter = 3;
    int32 key_count = 4;
}

message StorageEntry {
    string key = 1;
    string value = 2;
    string version_summary = 3; // String representation of Vector Clock
}